File: src/main/java/com/example/travelator/_8statictotoplevelfunction/version1/java/Shortlists.java
package com.example.travelator._8statictotoplevelfunction.version1.java;

import com.example.travelator._8statictotoplevelfunction.origin.HasPrice;
import com.example.travelator._8statictotoplevelfunction.origin.HasRating;

import java.util.Comparator;
import java.util.List;
import java.util.stream.Stream;

import static java.util.Comparator.comparing;

public class Shortlists {
	
	private Shortlists() {
	}
	
	public static <T> List<T> sorted(List<T> shortlist, Comparator<? super T> ordering) {
		return shortlist.stream()
			.sorted(ordering)
			.toList();
	}
	
	public static <T> List<T> removeItemAt(List<T> shortlist, Integer index) {
		return Stream.concat(
			shortlist.stream().limit(index), // index 만큼 스트림을 substream으로 만들어서 앞에 붙임
			shortlist.stream().skip((long) index + 1) // index + 1부터 스트림을 substream으로 만들어서 뒤에 붙임
		).toList();
	}
	
	public static Comparator<HasRating> byRating() {
		return Comparator.comparingDouble(HasRating::getRating).reversed();
	}
	
	public static Comparator<HasPrice> byPriceLowToHigh() {
		return comparing(HasPrice::getPrice);
	}
	
}

File: src/main/java/com/example/travelator/_8statictotoplevelfunction/version1/kotlin/Shortlists.kt
package com.example.travelator._8statictotoplevelfunction.version1.kotlin

import com.example.travelator._8statictotoplevelfunction.origin.HasPrice
import com.example.travelator._8statictotoplevelfunction.origin.HasRating
import java.util.stream.Stream

/**
 * 정적 메서드를 최상위 수준 함수로 변환을 했으나 최상위 클래스만을 생성할 뿐
 * @JvmStatic 까지는 만들지 못함
 */
object Shortlists {
    @JvmStatic
    fun <T> sorted(shortlist: List<T>, ordering: Comparator<in T>?): List<T> {
        return shortlist.stream()
            .sorted(ordering)
            .toList()
    }

    @JvmStatic
    fun <T> removeItemAt(shortlist: List<T>, index: Int): List<T> {
        return Stream.concat(
            shortlist.stream().limit(index.toLong()),  // index 만큼 스트림을 substream으로 만들어서 앞에 붙임
            shortlist.stream().skip(index.toLong() + 1) // index + 1부터 스트림을 substream으로 만들어서 뒤에 붙임
        ).toList()
    }

    @JvmStatic
    fun byRating(): Comparator<HasRating> {
        return Comparator.comparingDouble(HasRating::rating).reversed()
    }

    @JvmStatic
    fun byPriceLowToHigh(): Comparator<HasPrice> {
        return Comparator.comparing(HasPrice::price)
    }
}